File: desktop-client/pystudio/desktop_facade.py
from __future__ import annotations

from datetime import datetime
from pathlib import Path

from .git_service import GitResult, GitService
from .models import Project, ScriptFile
from .python_runner import ExecutionResult, InteractivePythonConsole, PythonRunner
from .signature_service import SignatureService


class DesktopFacade:
    """Facade: punto único de entrada para operaciones del IDE desktop."""

    def __init__(self) -> None:
        self.project: Project | None = None
        self.current_script: ScriptFile | None = None
        self.signatures = SignatureService()
        self.python_runner = PythonRunner()
        self.interactive_console = InteractivePythonConsole()
        self.git = GitService()

    def create_project(self, parent_dir: Path, project_name: str) -> Project:
        root = parent_dir / project_name
        root.mkdir(parents=True, exist_ok=True)
        project = Project(project_name, root)
        self.project = project
        project.save_manifest()
        self.git.init_repo(project.root)
        self.git.commit_all(project.root, self._auto_message("creacion de proyecto"))
        return project

    def open_project(self, root: Path) -> Project:
        self.project = Project.load(root)
        self.git.init_repo(self.project.root)
        return self.project

    def create_script(self, script_name: str) -> ScriptFile:
        project = self._require_project()
        if not script_name.endswith(".py"):
            script_name += ".py"
        path = project.root / script_name
        if path.exists():
            raise FileExistsError(f"El script {script_name} ya existe.")
        path.write_text("# Nuevo script PyStudio\n\nprint('Hola desde PyStudio')\n", encoding="utf-8")
        self.current_script = ScriptFile(script_name, path)
        project.save_manifest()
        self.git.commit_all(project.root, self._auto_message(f"crea {script_name}"))
        return self.current_script

    def load_script(self, script_name: str) -> ScriptFile:
        project = self._require_project()
        path = project.root / script_name
        if not path.exists():
            raise FileNotFoundError(f"No existe el archivo {script_name}.")
        self.current_script = ScriptFile(script_name, path)
        return self.current_script

    def list_scripts(self) -> list[ScriptFile]:
        project = self._require_project()
        return project.refresh_scripts()

    def save_current_script(self, content: str, commit: bool = True) -> GitResult | None:
        project = self._require_project()
        script = self._require_script()
        script.write_text(content)
        project.save_manifest()
        if commit:
            return self.git.commit_all(project.root, self._auto_message(f"guarda {script.name}"))
        return None

    def delete_current_script(self) -> GitResult:
        project = self._require_project()
        script = self._require_script()
        deleted_name = script.name
        script.path.unlink(missing_ok=True)
        self.current_script = None
        project.save_manifest()
        return self.git.commit_all(project.root, self._auto_message(f"borra {deleted_name}"))

    def sign_current_script(self, content: str) -> tuple[str, GitResult]:
        project = self._require_project()
        script = self._require_script()
        self.save_current_script(content, commit=False)
        signature = self.signatures.sign_file(script.path, project.signature_path)
        git_result = self.git.commit_all(project.root, self._auto_message(f"firma {script.name}"))
        return signature, git_result

    def verify_current_script(self, content: str) -> bool:
        project = self._require_project()
        script = self._require_script()
        self.save_current_script(content, commit=False)
        return self.signatures.verify_file(script.path, project.signature_path)

    def run_current_script(self, content: str) -> tuple[ExecutionResult, GitResult | None]:
        project = self._require_project()
        script = self._require_script()
        self.save_current_script(content, commit=False)
        result = self.python_runner.run_script(script.path)
        git_result = self.git.commit_all(project.root, self._auto_message(f"ejecuta {script.name}"))
        return result, git_result

    def run_interactive_line(self, line: str) -> ExecutionResult:
        return self.interactive_console.run_line(line)

    def git_history(self) -> str:
        project = self._require_project()
        return self.git.history(project.root)

    def commit_submission(self) -> GitResult:
        project = self._require_project()
        script = self._require_script()
        return self.git.commit_all(project.root, self._auto_message(f"entrega {script.name}"))

    def _require_project(self) -> Project:
        if not self.project:
            raise RuntimeError("Primero debe crear o abrir un proyecto.")
        return self.project

    def _require_script(self) -> ScriptFile:
        if not self.current_script:
            raise RuntimeError("Primero debe seleccionar o crear un script.")
        return self.current_script

    def _auto_message(self, action: str) -> str:
        return f"[auto] {action} {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}"

File: desktop-client/PyStudioDesktopSharp/Services/DesktopFacade.cs
using PyStudioDesktopSharp.Models;

namespace PyStudioDesktopSharp.Services;

public sealed class DesktopFacade
{
    public Project? Project { get; private set; }
    public ScriptFile? CurrentScript { get; private set; }

    private readonly SignatureService _signatures = new();
    private readonly PythonRunner _pythonRunner = new();
    private readonly GitService _git = new();

    public Project CreateProject(string parentDirectory, string projectName)
    {
        Project = Project.Create(parentDirectory, projectName);
        _git.InitRepo(Project.Root);
        _git.CommitAll(Project.Root, AutoMessage("creacion de proyecto"));
        return Project;
    }

    public Project OpenProject(string root)
    {
        Project = Project.Load(root);
        _git.InitRepo(Project.Root);
        return Project;
    }

    public ScriptFile CreateScript(string scriptName)
    {
        var project = RequireProject();
        scriptName = scriptName.Trim();
        if (!scriptName.EndsWith(".py", StringComparison.OrdinalIgnoreCase))
            scriptName += ".py";

        string path = Path.Combine(project.Root, scriptName);
        if (File.Exists(path))
            throw new IOException($"El script {scriptName} ya existe.");

        File.WriteAllText(path, "# Nuevo script PyStudio\n\nprint('Hola desde PyStudio')\n");
        CurrentScript = new ScriptFile(scriptName, path);
        project.SaveManifest();
        _git.CommitAll(project.Root, AutoMessage($"crea {scriptName}"));
        return CurrentScript;
    }

    public ScriptFile LoadScript(string scriptName)
    {
        var project = RequireProject();
        string path = Path.Combine(project.Root, scriptName);
        if (!File.Exists(path))
            throw new FileNotFoundException($"No existe el archivo {scriptName}.");

        CurrentScript = new ScriptFile(scriptName, path);
        return CurrentScript;
    }

    public List<ScriptFile> ListScripts()
    {
        return RequireProject().RefreshScripts();
    }

    public GitResult? SaveCurrentScript(string content, bool commit = true)
    {
        var project = RequireProject();
        var script = RequireScript();
        script.WriteText(content);
        project.SaveManifest();
        return commit ? _git.CommitAll(project.Root, AutoMessage($"guarda {script.Name}")) : null;
    }

    public GitResult DeleteCurrentScript()
    {
        var project = RequireProject();
        var script = RequireScript();
        string deletedName = script.Name;
        if (File.Exists(script.Path)) File.Delete(script.Path);
        CurrentScript = null;
        project.SaveManifest();
        return _git.CommitAll(project.Root, AutoMessage($"borra {deletedName}"));
    }

    public (string Signature, GitResult GitResult) SignCurrentScript(string content)
    {
        var project = RequireProject();
        var script = RequireScript();
        SaveCurrentScript(content, commit: false);
        string signature = _signatures.SignFile(script.Path, project.SignaturePath);
        GitResult gitResult = _git.CommitAll(project.Root, AutoMessage($"firma {script.Name}"));
        return (signature, gitResult);
    }

    public bool VerifyCurrentScript(string content)
    {
        var project = RequireProject();
        var script = RequireScript();
        SaveCurrentScript(content, commit: false);
        return _signatures.VerifyFile(script.Path, project.SignaturePath);
    }

    public async Task<(ExecutionResult Result, GitResult? GitResult)> RunCurrentScriptAsync(string content)
    {
        var project = RequireProject();
        var script = RequireScript();
        SaveCurrentScript(content, commit: false);
        ExecutionResult result = await _pythonRunner.RunScriptAsync(script.Path);
        GitResult gitResult = _git.CommitAll(project.Root, AutoMessage($"ejecuta {script.Name}"));
        return (result, gitResult);
    }

    public async Task<ExecutionResult> RunInteractiveLineAsync(string line)
    {
        return await _pythonRunner.RunInteractiveLineAsync(line);
    }

    public string GitHistory()
    {
        return _git.History(RequireProject().Root);
    }

    public GitResult CommitSubmission()
    {
        var project = RequireProject();
        var script = RequireScript();
        return _git.CommitAll(project.Root, AutoMessage($"entrega {script.Name}"));
    }

    private Project RequireProject()
    {
        return Project ?? throw new InvalidOperationException("Primero debe crear o abrir un proyecto.");
    }

    private ScriptFile RequireScript()
    {
        return CurrentScript ?? throw new InvalidOperationException("Primero debe seleccionar o crear un script.");
    }

    private static string AutoMessage(string action)
    {
        return $"[auto] {action} {DateTime.Now:yyyy-MM-dd HH:mm:ss}";
    }
}
